lib/core/Axios.js

```diff
@@ -45,20 +45,61 @@ Axios.prototype.request = function request(config) {
     config.method = 'get';
   }
 
-  // Hook up interceptors middleware
-  var chain = [dispatchRequest, undefined];
-  var promise = Promise.resolve(config);
-
+  // filter out skipped interceptors
+  var requestInterceptorChain = [];
+  var synchronousRequestInterceptors = true;
   this.interceptors.request.forEach(function unshiftRequestInterceptors(interceptor) {
-    chain.unshift(interceptor.fulfilled, interceptor.rejected);
+    if (typeof interceptor.runWhen === 'function' && interceptor.runWhen(config) === false) {
+      return;
+    }
+
+    synchronousRequestInterceptors = synchronousRequestInterceptors && interceptor.synchronous;
+
+    requestInterceptorChain.unshift(interceptor.fulfilled, interceptor.rejected);
   });
 
+  var responseInterceptorChain = [];
   this.interceptors.response.forEach(function pushResponseInterceptors(interceptor) {
-    chain.push(interceptor.fulfilled, interceptor.rejected);
+    responseInterceptorChain.push(interceptor.fulfilled, interceptor.rejected);
   });
 
-  while (chain.length) {
-    promise = promise.then(chain.shift(), chain.shift());
+  var promise;
+
+  if (!synchronousRequestInterceptors) {
+    var chain = [dispatchRequest, undefined];
+
+    Array.prototype.unshift.apply(chain, requestInterceptorChain);
+    chain.concat(responseInterceptorChain);
+
+    promise = Promise.resolve(config);
+    while (chain.length) {
+      promise = promise.then(chain.shift(), chain.shift());
+    }
+
+    return promise;
+  }
+
+
+  var newConfig = config;
+  while (requestInterceptorChain.length) {
+    var onFulfilled = requestInterceptorChain.shift();
+    var onRejected = requestInterceptorChain.shift();
+    try {
+      newConfig = onFulfilled(newConfig);
+    } catch (error) {
+      onRejected(error);
```

Comment:
If `onRejected` had solved the error, we should continue the loop. If not, should drop to the next `onRejected`.

Reply:
Hmm, this would make the functionality different from the `asynchronous` behavior then. 

When we have async interceptors we do a promise chain: 

`promise = promise.then(chain.shift(), chain.shift());`

When any interceptor throws in a chained promise, the whole thing errors out.

Reply:
@SashaKoro Sorry for my long long latency. I mean,

```
var f = function(name) {
    return function(...args) {
        console.log(name, args)
    }
}

Promise.reject('bad')
     .then(f('f1'), f('f2'))
     .then(f('f3'), f('f4'))

// outputs
f2 [ 'bad' ]
f3 [ undefined ]
```

Even though the chain has an error, but `f2` can resolve it and make the loop go on.